### PR TITLE
Feature-96 MQTT support Implemented.

### PR DESCRIPTION
**Pull request by amitrnec**
MQTT support pull request

### CODE REVIEW

File: Core/SubscriptionManager/src/main/resources/application.yml [hunk 1]
```diff
@@ -22,7 +22,10 @@ csource:
     topic: CSOURCE_REGISTRATION
   notification:
     topic: CONTEXT_SOURCE_NOTIFICATION
-
+query:
```
Comment (naveennec) [inline Core/SubscriptionManager/src/main/resources/application.yml:25]: [Minor]indentation is not proper. it should be same as notification(above one) level.

File: Core/SubscriptionManager/src/main/java/eu/neclab/ngsildbroker/subscriptionmanager/service/BaseNotificationHandler.java [hunk 2]
```diff
@@ -578,13 +579,20 @@ public Part getPart(String name) throws IOException, ServletException {
...
+            
 			dataArray.add(jsonTree.get("data"));
 			((ObjectNode) jsonTree).set("data", dataArray);
-
+			Map<String,String> map = new HashMap<String,String>();
+			if(context!=null && !context.isEmpty()) {
+			map.put(NGSIConstants.ACCEPTED_LINK, context.get(0).toString());
+			} 
+			map.put(NGSIConstants.CONTENT_TYPE, acceptHeader);
+            Map<String,Object> mapObj = new HashMap<String, Object>();
```
Comment (naveennec) [inline Core/SubscriptionManager/src/main/java/eu/neclab/ngsildbroker/subscriptionmanager/service/BaseNotificationHandler.java:590]: [minor]Indentation is not proper

File: Commons/src/main/java/eu/neclab/ngsildbroker/commons/constants/NGSIConstants.java [hunk 3]
```diff
@@ -180,4 +180,15 @@
 	public static final String MQTT_VERSION = "mqtt_version";
 	public static final String DEFAULT_DATA_SET_ID = "https://uri.etsi.org/ngsi-ld/default-data-set-id";
 	public final static String QUERY_PARAMETER_DELETE_ALL  = "deleteAll";
+	public static final String NGSI_LD_NOTIFIERINFO = "https://uri.etsi.org/ngsi-ld/default-context/notifierinfo";
+	public static final String NGSI_LD_MQTT_QOS = "https://uri.etsi.org/ngsi-ld/default-context/qos";
+	public static final String NGSI_LD_MQTT_VERSION = "https://uri.etsi.org/ngsi-ld/default-context/version";
+	public static final String DEFAULT_MQTT_QOS = "0";
+	public static final String DEFAULT_MQTT_VERSION = "5";
```
Comment (naveennec) [inline Commons/src/main/java/eu/neclab/ngsildbroker/commons/constants/NGSIConstants.java:187]: [Major]Default version of MQTT should be "mqtt5.0" not 5.

File: Commons/src/main/java/eu/neclab/ngsildbroker/commons/constants/NGSIConstants.java [hunk 4]
```diff
@@ -180,4 +180,15 @@
 	public static final String MQTT_VERSION = "mqtt_version";
 	public static final String DEFAULT_DATA_SET_ID = "https://uri.etsi.org/ngsi-ld/default-data-set-id";
 	public final static String QUERY_PARAMETER_DELETE_ALL  = "deleteAll";
+	public static final String NGSI_LD_NOTIFIERINFO = "https://uri.etsi.org/ngsi-ld/default-context/notifierinfo";
+	public static final String NGSI_LD_MQTT_QOS = "https://uri.etsi.org/ngsi-ld/default-context/qos";
+	public static final String NGSI_LD_MQTT_VERSION = "https://uri.etsi.org/ngsi-ld/default-context/version";
+	public static final String DEFAULT_MQTT_QOS = "0";
```
Comment (naveennec) [inline Commons/src/main/java/eu/neclab/ngsildbroker/commons/constants/NGSIConstants.java:186]: [Suggestion] QOS is integer value why we are taking it as string as possible value of QOS is 0,1,2 so better to take it as Integer.

File: Commons/src/main/java/eu/neclab/ngsildbroker/commons/serialization/SubscriptionGsonAdapter.java [hunk 5]
```diff
@@ -143,6 +145,21 @@ public Subscription deserialize(JsonElement json, Type typeOfT, JsonDeserializat
 				try {
 					endPoint.setUri(new URI(jsonEndPoint.getAsJsonArray(NGSIConstants.NGSI_LD_URI).get(0)
 							.getAsJsonObject().get(NGSIConstants.JSON_LD_VALUE).getAsString()));
+					Map<String,String> infoSettingNotifier = new HashMap<String,String>();
```
Comment (naveennec) [inline Commons/src/main/java/eu/neclab/ngsildbroker/commons/serialization/SubscriptionGsonAdapter.java:148]: [Trival] Please write comment in the code for better understanding.

File: Commons/src/main/java/eu/neclab/ngsildbroker/commons/serialization/SubscriptionGsonAdapter.java [hunk 6]
```diff
@@ -143,6 +145,21 @@ public Subscription deserialize(JsonElement json, Type typeOfT, JsonDeserializat
...
+					if (jsonEndPoint.has(NGSIConstants.NGSI_LD_NOTIFIERINFO)
+							&& jsonEndPoint.get(NGSIConstants.NGSI_LD_NOTIFIERINFO).isJsonArray()) {
+						JsonObject info = jsonEndPoint.getAsJsonArray(NGSIConstants.NGSI_LD_NOTIFIERINFO).get(0).getAsJsonObject();
+						String mqttQos = info.getAsJsonArray(NGSIConstants.NGSI_LD_MQTT_QOS).get(0).getAsJsonObject().get(NGSIConstants.JSON_LD_VALUE).getAsString();
+						String mqttVersion = info.getAsJsonArray(NGSIConstants.NGSI_LD_MQTT_VERSION).get(0).getAsJsonObject().get(NGSIConstants.JSON_LD_VALUE).getAsString();
+						infoSettingNotifier.put(NGSIConstants.MQTT_QOS, mqttQos);
+						infoSettingNotifier.put(NGSIConstants.MQTT_VERSION, mqttVersion);
+						endPoint.setNotifierInfo(infoSettingNotifier);
+					} //else {
```
Comment (naveennec) [inline Commons/src/main/java/eu/neclab/ngsildbroker/commons/serialization/SubscriptionGsonAdapter.java:157]: [Trival] please remove the commented section if this code is of no use.

File: Commons/src/main/java/eu/neclab/ngsildbroker/commons/serialization/SubscriptionGsonAdapter.java [hunk 7]
```diff
@@ -317,6 +334,33 @@ public JsonElement serialize(Subscription src, Type typeOfSrc, JsonSerialization
 							context.serialize(notification.getEndPoint().getUri().toString()));
 					tempArray.add(tempObj);
 					endPoint.add(NGSIConstants.NGSI_LD_URI, tempArray);
+					//endPointArray.add(endPoint);
+					//notificationObj.add(NGSIConstants.NGSI_LD_ENDPOINT, endPointArray);
+				}
+				if (notification.getEndPoint().getNotifierInfo() != null) {
```
Comment (naveennec) [inline Commons/src/main/java/eu/neclab/ngsildbroker/commons/serialization/SubscriptionGsonAdapter.java:340]: [Trival] Please write comment in the code for better understanding.

File: Commons/src/main/java/eu/neclab/ngsildbroker/commons/serialization/SubscriptionGsonAdapter.java [hunk 8]
```diff
@@ -317,6 +334,33 @@ public JsonElement serialize(Subscription src, Type typeOfSrc, JsonSerialization
...
+					
+					JsonObject notifierEndPoint = new JsonObject();
+					JsonArray notifierEndPointArray = new JsonArray();
+					 if(notification.getEndPoint().getNotifierInfo().get(NGSIConstants.MQTT_QOS) != null) {
```
Comment (naveennec) [inline Commons/src/main/java/eu/neclab/ngsildbroker/commons/serialization/SubscriptionGsonAdapter.java:344]: [minor] indentation is not proper. Please correct it.

File: Commons/src/main/java/eu/neclab/ngsildbroker/commons/serialization/SubscriptionGsonAdapter.java [hunk 9]
```diff
@@ -8,7 +8,9 @@
 import java.time.temporal.TemporalAccessor;
 import java.util.ArrayList;
 import java.util.Date;
+import java.util.HashMap;
 import java.util.Iterator;
+import java.util.Map;
 import java.util.Map.Entry;
```
Comment (naveennec) [inline Commons/src/main/java/eu/neclab/ngsildbroker/commons/serialization/SubscriptionGsonAdapter.java:14]: [Minor] when Map is already imported then Map.Entry should be removed.

File: Core/SubscriptionManager/src/main/java/eu/neclab/ngsildbroker/subscriptionmanager/service/NotificationHandlerMQTT.java [hunk 10]
```diff
@@ -58,7 +64,14 @@ private MqttClient getClient(URI callback, Map<String, String> clientSettings) {
...
+			if(clientSettings != null) {
+				mqttVersion = clientSettings.get(NGSIConstants.MQTT_VERSION);
+				
+			} else {
+				mqttVersion = NGSIConstants.DEFAULT_MQTT_VERSION; 
+			}
+			
 			int port = callback.getPort();
 			if (port == -1) {
 				port = 1883;
```
Comment (naveennec) [inline Core/SubscriptionManager/src/main/java/eu/neclab/ngsildbroker/subscriptionmanager/service/NotificationHandlerMQTT.java:77]: [Query] the default MQTT port is 1883 for MQTT over TCP and 8883 for mqtts. Are we taking care of MQTTS

**Review by ScorpioBroker — APPROVED** [review]
LGTM

**Review by naveennec — CHANGES_REQUESTED** [review]

**Pull request merged**